app/src/main/java/org/mozilla/fenix/gleanplumb/MessageNotificationWorker.kt

```diff
@@ -57,46 +60,81 @@ class MessageNotificationWorker(
 
-            // Generate the processed Message action
-            val processedAction = nimbusMessagingController.processMessageActionToUri(nextMessage)
-            val actionIntent = Intent(Intent.ACTION_VIEW, processedAction)
-
             NotificationManagerCompat.from(context).notify(
                 MESSAGE_TAG,
-                SharedIdsHelper.getNextIdForTag(context, nextMessage.id),
-                buildNotification(nextMessage, actionIntent),
+                SharedIdsHelper.getNextIdForTag(context, messageAsDisplayed.id),
+                buildNotification(
+                    context,
+                    nimbusMessagingController,
+                    messageAsDisplayed,
+                ),
             )
         }
 
         return Result.success()
     }
 
-    private fun Message.shouldDisplayMessage() = metadata.displayCount == 0
+    private fun buildNotification(
+        context: Context,
+        nimbusMessagingController: NimbusMessagingController,
+        message: Message,
+    ): Notification {
+        val onClickPendingIntent =
+            createOnClickPendingIntent(context, nimbusMessagingController, message)
+        val onDismissPendingIntent = createOnDismissPendingIntent(context, message)
+
+        return createBaseNotification(
+            context,
+            MARKETING_CHANNEL_ID,
+            message.data.title,
+            message.data.text,
+            onClickPendingIntent,
+            onDismissPendingIntent,
+        )
+    }
 
-    private fun buildNotification(message: Message, intent: Intent): Notification {
-        with(applicationContext) {
-            val pendingIntent = PendingIntent.getActivity(
-                this,
-                SharedIdsHelper.getNextIdForTag(this, NOTIFICATION_PENDING_INTENT_TAG),
-                intent,
-                IntentUtils.defaultIntentPendingFlags,
-            )
+    private fun createOnDismissPendingIntent(
+        context: Context,
+        message: Message,
+    ): PendingIntent {
+        val intent = Intent(context, NotificationDismissedService::class.java)
+        intent.putExtra(MESSAGE_METADATA, message.metadata.toJson())
+
+        return PendingIntent.getService(
+            context,
+            SharedIdsHelper.getNextIdForTag(context, NOTIFICATION_PENDING_INTENT_TAG),
+            intent,
+            IntentUtils.defaultIntentPendingFlags,
+        )
+    }
 
-            return createBaseNotification(
-                this,
-                MARKETING_CHANNEL_ID,
-                message.data.title,
-                message.data.text,
-                pendingIntent,
-            )
-        }
+    private fun createOnClickPendingIntent(
+        context: Context,
+        nimbusMessagingController: NimbusMessagingController,
+        message: Message,
+    ): PendingIntent {
+        val intent = nimbusMessagingController.processMessageActionToIntent(message)
```

Comment:
IIUC, the action's UUID is being generated and recorded with Glean as if the message was clicked on.

Reply:
good spot! 🙌